Addition method testing with JUnit

julianmateu opened the pull request:
Is it OK to use assertTrue?, if not:
-Should I override the Equals method or define a new assertEquals?
-How do I override the Equals(Object object) method to use assertEquals in JUnit?
     http://stackoverflow.com/questions/6060848/junit-assertequals-fails-for-two-objects
     http://stackoverflow.com/questions/2316220/java-override-object-equals-method

julianmateu committed:
Added setters, getters and equals()

In Quaternion.java:
- Add setters for r,i,j and k.
- Add getters for r,i,j and k.
- Add a method equals() to assert equality between two Quaternions.

julianmateu committed:
Modified Documentation

julianmateu committed:
Add Testing files

-Add TestQuaternion.java with methods to test Equals() and Add() methods from the Quaternion class.
-Add script test.sh to run the tests.

(Is it OK to use assertTrue?, if not:)
(-Should I override the Equals method or define a new assertEquals?)
(-How do I override the Equals(Object object) method to use assertEquals in JUnit?)

basicNew commented on src/Quaternion.java:
```diff
@@ -31,6 +68,19 @@ public Quaternion add(Quaternion addend) {
+     */
+    //@Override
+    public boolean equals(Quaternion compared) {
+        
```
should use getters and setters now that you have them :)

basicNew commented on src/Quaternion.java:
```diff
@@ -31,6 +68,19 @@ public Quaternion add(Quaternion addend) {
+        if (this.i != compared.i) return false;
+        if (this.j != compared.j) return false;
+        if (this.k != compared.k) return false;
+        else return true;
```
Mostly a matter of style, but why not a single return using `AND`s?

basicNew commented on src/TestQuaternion.java:
```diff
@@ -0,0 +1,22 @@
+    public void testEquals() {
+        Quaternion a = new Quaternion(1,0,-1,5);
+
+        assertTrue(a.equals(new Quaternion(1,0,-1,5)));
```
Maybe `assertEquals(...,...)` is better here to convey the idea

julianmateu replied:
That was one of the questions I had, because I tried with `assertEquals(a,new Quaternion(1,0,-1,5));` and got a message saying:

```
There was 1 failure:
1) testEquals(TestQuaternion)junit.framework.AssertionFailedError: expected:<( 1.0 ; 0.0 ; -1.0 ; 5.0 )> but was:<( 1.0 ; 0.0 ; -1.0 ; 5.0 )>
```

I did some research (http://stackoverflow.com/questions/6060848/junit-assertequals-fails-for-two-objects), and found that when you define a method called `equals()` for a class you create, you should write one with the following header, because this is the one JUnit uses in `assertEquals()`:

```
@Override
public boolean equals(Object other)
```

However, when I tried this I came across the problem that a generic Object doesn't have the fields (r,i,j,k) nor the setters and getters.

I searched how to define an `equals(Object)` method that overrides the default one and found (http://stackoverflow.com/questions/2316220/java-override-object-equals-method) that you can do something like check that the object is the type you want and then cast it, but I got a little confused by the comments, and since I was certain that `assertTrue()` worked, I used that instead of `assertEquals()`.

basicNew replied:
That's great, no problem! Let's discuss this next time we meet.

On Tue, Jan 26, 2016 at 9:53 AM, Julián Mateu notifications@github.com
wrote:

> In src/TestQuaternion.java
> https://github.com/julianmateu/Quaternion/pull/1#discussion_r50830676:
> 
> > @@ -0,0 +1,22 @@
> > +import junit.framework.*;
> > +
> > +public class TestQuaternion extends TestCase {
> > +
> > -    public TestQuaternion(String name) {
> > -        super(name);
> > -    }
> >   +
> > -    public void testEquals() {
> > -        Quaternion a = new Quaternion(1,0,-1,5);
> >   +
> > -        assertTrue(a.equals(new Quaternion(1,0,-1,5)));
> 
> That was one of the questions I had, because I tried with assertEquals(a,new
> Quaternion(1,0,-1,5)); and got a message saying:
> 
> There was 1 failure:
> 1) testEquals(TestQuaternion)junit.framework.AssertionFailedError: expected:<( 1.0 ; 0.0 ; -1.0 ; 5.0 )> but was:<( 1.0 ; 0.0 ; -1.0 ; 5.0 )>
> 
> I did some research (
> http://stackoverflow.com/questions/6060848/junit-assertequals-fails-for-two-objects),
> and found that when you define a method called equals() for a class you
> create, you should write one with the following header, because this is the
> one JUnit uses in assertEquals():
> 
> @Override
> public boolean equals(Object other)
> 
> However, when I tried this I came across the problem that a generic Object
> doesn't have the fields (r,i,j,k) nor the setters and getters.
> 
> I searched how to define an equals(Object) method that overrides the
> default one and found (
> http://stackoverflow.com/questions/2316220/java-override-object-equals-method)
> that you can do something like check that the object is the type you want
> and then cast it, but I got a little confused by the comments, and because
> I was certain that assertTrue() worked, I used that.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/julianmateu/Quaternion/pull/1/files#r50830676.

basicNew commented on src/TestQuaternion.java:
```diff
@@ -0,0 +1,22 @@
+        Quaternion b = new Quaternion(1,0,-1,-1);
+        Quaternion result = new Quaternion(2,1,0,-2);
+
+        assertTrue(result.equals(a.add(b)));
```
Again, something along the lines of `assertEquals(a.add(b), result)` is more readable and less code :)

basicNew commented:
Looks good for a starter. The question now is if you can come up with any edge cases to add to your tests :)

julianmateu committed:
Tried using assertEquals() but didn't work

In TestQuaternion.java:
-Used the assertEquals() method in the testEquals() method, but got an error, probably due to not having defined an equals(Object) method.

julianmateu committed:
Modified equals() and used assertEquals()

In Quaternion.java:
-Modified equals() method using getters and AND's.
-Defined an equals(Object) method.

The assertEquals now worked!

julianmateu committed:
Used assertEquals() in testAddition()

julianmateu committed:
Added test from file

Added a TestQuaternionDataFile test, to run test cases from files.
Added a file named testdata.txt with some tests I could think about.

julianmateu committed:
Deleted testing from data file

After the meeting of 2016/01/27, deleted the Testing form data file, and moved the test cases to the main Test (TestQuaternion.java).
